service/datasync/exec_data_sync.go

```diff
@@ -0,0 +1,238 @@
+	err := e.getExecData()
+	if errors.Is(err, storage.ErrObjectNotExist) {
+		e.log.Debug().Msg("next execution record not available, download stopped")
+		return
+	}
+	if err != nil {
+		e.log.Error().Err(err).Msg("could not download execution records")
+		return
+	}
+}
+
+func (e *ExecDataSync) getExecData() error {
+
+	for {
+
+		// We only want to retrieve and process execData blocks until the buffer is full.
+		if uint(e.buffer.Len()) >= e.limit {
+			e.log.Debug().Uint("limit", e.limit).Msg("buffer full, stopping execution record pull")
+			return nil
+		}
+
+		// We only want to retrieve and process files for blocks that have already
+		// been finalized, in the order that they have been finalized. This
+		// causes some latency, as we don't download until after a block is
+		// finalized, even if the data is available before. However, it seems to
+		// be the only way to make sure trie updates are delivered to the mapper
+		// in the right order without changing the way uploads work.
+		if uint(e.queue.Len()) == 0 {
+			e.log.Debug().Msg("queue empty, stopping execution record download")
+			return nil
+		}
+
+		// Get the name of the file based on the block ID. The file n
+		blockID := e.queue.PopBack().(flow.Identifier)
+		record, err := e.pullData(context.Background(), blockID)
+		if err != nil {
+			e.queue.PushBack(blockID)
+			return fmt.Errorf("could not pull execution record (name: %s): %w", blockID, err)
+		}
+
+		e.log.Debug().
+			Hex("blockID", blockID[:]).
+			Msg("pushing execution record into buffer")
+
+		e.buffer.PushFront(record)
+	}
+}
+
+func (e *ExecDataSync) pullData(ctx context.Context, blockID flow.Identifier) (*uploader.BlockData, error) {
+	// query AN from cached connection
+	req := &execData.GetExecutionDataByBlockIDRequest{BlockId: blockID[:]}
+	res, err := e.execDataApi.GetExecutionDataByBlockID(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return convertToBlockData(res.BlockExecutionData)
+}
+
+func convertToBlockData(data *entities.BlockExecutionData) (*uploader.BlockData, error) {
+	// convert to block exec data defined in flow-go
+	bed, err := convert.MessageToBlockExecutionData(data, flow.Mainnet.Chain())
+	if err != nil {
+		return nil, err
+	}
+
+	collections := make([]*entity.CompleteCollection, 0)
+	events := make([]*flow.Event, 0)
+	trieUpdates := make([]*ledger.TrieUpdate, 0)
+
+	for _, ced := range bed.ChunkExecutionDatas {
+		// combine the collections in the chunks
+		collectionToAdd := entity.CompleteCollection{
+			Guarantee:    nil,
+			Transactions: ced.Collection.Transactions,
+		}
+		collections = append(collections, &collectionToAdd)
+
+		// combine the events in the chunks
+		eventsToAdd := make([]*flow.Event, len(ced.Events))
+		for i := range ced.Events {
+			eventsToAdd[i] = &ced.Events[i]
+		}
+		events = append(events, eventsToAdd...)
+
+		// combine the trie updates in the chunks
+		trieUpdates = append(trieUpdates, ced.TrieUpdate)
+	}
+	blockData := uploader.BlockData{
```

Comment:
Lets add some docs here to explain why some of the fields are assigned `nil` (to decrease overlap)